File: src/motion_py_pkg/motion_py_pkg/manipulator/verify.py
import numpy as np
from .forward import forward
from .jacobian import jacobian


def verify(path_angles, MDH_base, print_transforms=False, robot="Robot"):
    """
    Verify the inverse kinematics solution and check for singularities along the path.

    Parameters:
        path_angles     (np.ndarray): 5 x N array of joint angles (radians) for each path point.
        MDH_base        (np.ndarray): 6x5 MDH table (with symbolic theta positions — angles
                                      are substituted per path point inside this function).
        print_transforms (bool):      If True, print T06 at every path point.
        robot           (str):        Robot name for display output.
    """
    n_points = path_angles.shape[1]
    singular = False
    close    = False

    for i in range(n_points):
        th = path_angles[:, i]          # [TH1 … TH5]

        # Build numeric MDH for this configuration
        MDH_num = MDH_base.copy().astype(float)
        MDH_num[0, 3] = th[0]
        MDH_num[1, 3] = th[1]
        MDH_num[2, 3] = th[2]
        MDH_num[3, 3] = th[3]
        MDH_num[4, 3] = th[4]
        # Row 5 (joint 6) has a fixed angle of 0 — already set in the base table.

        if print_transforms:
            _, _, _, _, _, _, T06 = forward(MDH_num) if False else (None,)*7  # placeholder
            # Re-use jacobian's internal forward call below; compute T06 separately
            from forward import forward as fwd
            T06 = fwd(MDH_num)[-1]
            print(f"Transformation Matrix T for the {robot} at target {i + 1}:")
            print(T06, "\n")

        J, LinJ = jacobian(MDH_num)

        r_lin  = np.linalg.matrix_rank(LinJ)
        r_full = np.linalg.matrix_rank(J)
        s_lin  = np.linalg.svd(LinJ, compute_uv=False)
        s_full = np.linalg.svd(J,    compute_uv=False)

        if r_lin < 3 or r_full < 5:
            print(f"Determinant of the {robot} is zero at target {i + 1}, singular configuration.")
            singular = True
        elif np.min(s_lin) < 1e-4 or np.min(s_full) < 1e-4:
            print(f"Jacobian is close to singular at target {i + 1} for the {robot}.")
            close = True

    if singular:
        print("\nYour path is not safe as you are at a singularity at some point.\n")
    elif close:
        print("\nYour path may not be safe as you are close to singularity.\n")
    else:
        print("\nYour path is completely safe.\n")

File: src/manipulator_matlab/Verify.m
function Verify(Path, T06, J, LinJ, print, robot)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function Verify(Path, T06, Det, print, robot)
% 
% Task: Verify the Inverse Kinematics and if we have singularities or are close to one
%
% Inputs:
%	- Path: Matrix containing all of the path's targets
%   - T06: The final transformation matrix of the robot
%   - Det: The determinant of the Jacobian of the robot
%   - print: boolean if you want to print the results or not
%   - robot: the robot model
%
% author: Sébastien Dubois, s.j.dubois.116@cranfield.ac.uk
% date: 12/12/2025
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Calculate the transformation matrix with the angles found in the Path
if print
    for i = 1:size(Path, 2)
        T = double(T06(Path(1, i), Path(2, i), Path(3, i), Path(4, i), Path(5, i)));    % Calculate the transformation matrix with the current angle set to verify with our calculations
        fprintf('Transformation Matrix T for the %s at target %d:\n', robot, i);
        disp(T);
    end
end

%% Calculate the determinant with the angle configurations in the Path
Sing = false;
Close = false;
for i = 1:size(Path, 2)
    r = rank(double(LinJ(Path(1, i), Path(2, i), Path(3, i), Path(4, i), Path(5, i))));
    s = svd(double(LinJ(Path(1, i), Path(2, i), Path(3, i), Path(4, i), Path(5, i))));
    rFull = rank(double(J(Path(1, i), Path(2, i), Path(3, i), Path(4, i), Path(5, i))));
    sFull = svd(double(LinJ(Path(1, i), Path(2, i), Path(3, i), Path(4, i), Path(5, i))));
    if  r < 3 || rFull < 5
        fprintf('Determinant of the %s is zero at target %d, singular configuration.\n', robot, i);   % Print the target number if you are at a singularity
        Sing = true;
    elseif min(s) < 1e-4 || min(sFull) < 1e-4
        fprintf('Determinant = %d is close to zero at target %d, potential singular configuration for the %s.\n', D, i, robot); % Print the target number if you are close to singularity
        Close = true;
    end
end

%% Print necessary informations about the path
if Sing
    fprintf('\nYour path is not safe as you are at a singularity at some point\n\n');
elseif Close
    fprintf('\nYour path may not be safe as you are close to singularity\n\n');
else
    fprintf('\nYour path is completely safe\n\n')
end